patroni/quorum.py

```diff
@@ -0,0 +1,431 @@
+                yield from self.quorum_update(len(self.sync) - self.numsync, self.voters)
+
+    def __remove_gone_nodes(self) -> Iterator[Transition]:
+        """Remove inactive nodes from ``synchronous_standby_names`` and from ``/sync`` key.
+
+        :yields: transitions as :class:`Transition` objects.
+        """
+        to_remove = self.sync - self.active
+        if to_remove and self.sync == to_remove:
+            logger.debug("Removing nodes: %s", to_remove)
+            yield from self.quorum_update(0, CaseInsensitiveSet(), adjust_quorum=False)
+            yield from self.sync_update(0, CaseInsensitiveSet())
+        elif to_remove:
+            logger.debug("Removing nodes: %s", to_remove)
+            can_reduce_quorum_by = self.quorum
+            # If we can reduce quorum size try to do so first
+            if can_reduce_quorum_by:
+                # Pick nodes to remove by sorted order to provide deterministic behavior for tests
+                remove = CaseInsensitiveSet(sorted(to_remove, reverse=True)[:can_reduce_quorum_by])
+                sync = CaseInsensitiveSet(self.sync - remove)
+                # when removing nodes from sync we can safely increase numsync if requested
+                numsync = min(self.sync_wanted, len(sync)) if self.sync_wanted > self.numsync else self.numsync
+                yield from self.sync_update(numsync, sync)
+                voters = CaseInsensitiveSet(self.voters - remove)
+                to_remove &= self.sync
+                yield from self.quorum_update(len(voters) - self.numsync, voters,
+                                              adjust_quorum=not to_remove)
+            if to_remove:
+                assert self.quorum == 0
+                numsync = self.numsync - len(to_remove)
+                sync = CaseInsensitiveSet(self.sync - to_remove)
+                voters = CaseInsensitiveSet(self.voters - to_remove)
+                sync_decrease = numsync - min(self.sync_wanted, len(sync))
+                quorum = min(sync_decrease, len(voters) - 1) if sync_decrease else 0
+                yield from self.quorum_update(quorum, voters, adjust_quorum=False)
+                yield from self.sync_update(numsync, sync)
+
+    def __add_new_nodes(self) -> Iterator[Transition]:
+        """Add new active nodes to ``synchronous_standby_names`` and to ``/sync`` key.
+
+        :yields: transitions as :class:`Transition` objects.
+        """
+        to_add = self.active - self.sync
+        if to_add:
+            # First get to requested replication factor
+            logger.debug("Adding nodes: %s", to_add)
+            sync_wanted = min(self.sync_wanted, len(self.sync | to_add))
+            increase_numsync_by = sync_wanted - self.numsync
+            if increase_numsync_by > 0:
+                if self.sync:
+                    add = CaseInsensitiveSet(sorted(to_add)[:increase_numsync_by])
+                    increase_numsync_by = len(add)
```

Comment:
We need this because `to_add` maybe have less members that `increase_numsync_by` initially wants, right?

Reply:
Its a corner case / optimization.
Adding nodes and changing the replication factor are two different operations and supposed to be done one after another, which is not always optimal.

Example:
- `/sync`: `{leader: a, sync_standby: b, quorum: 0}`
- `synchronous_standby_names`: `ANY 1 {b}`

We want to add node `c` and increase the replication factor from 1 to 2.
Non optimal way of doing that would be:
1. `/sync`: `{leader: a, sync_standby: b,c, quorum: 1}` (add node /sync key with quorum increase)
2. `synchronous_standby_names`: `ANY 1 {b,c}` (add node no ssn)
3. `synchronous_standby_names`: `ANY 2 {b,c}` (increase replication factor to requested value)
4. `/sync`: `{leader: a, sync_standby: b,c, quorum: 0}` (reduce quorum)
`2` and `3` will be merged by `__iter__` method.

With optimization (because we know that the replication factor should be increased anyway) it produces the following:
1. `synchronous_standby_names`: `ANY 2 {b,c}` (add node no ssn and increase the replication factor)
2. `/sync`: `{leader: a, sync_standby: b,c, quorum: 0}` (add node to /sync key).

Besides that there could be other corner cases. When for example we have only 2 replicas (one is already in ssn and we are adding the second one), but the requested replication factor is 3 or higher.